pkg/controller/persistentvolume/persistentvolume_provisioner_controller.go

```diff
@@ -0,0 +1,314 @@
+		glog.V(5).Infof("PersistentVolumeClaim[%s/%s] missing provisionable annotation from binder.  No volume will be created.", pvc.Namespace, pvc.Name)
+		return nil
+	}
+
+	if _, exists := pvc.Annotations[provisionedKey]; exists {
+		glog.V(5).Infof("PersistentVolumeClaim[%s/%s] has already been provisioned.  No volume will be creatd.", pvc.Namespace, pvc.Name)
+		return nil
+	}
+
+	if _, exists := provisionedClaims[ClaimToProvisionableKey(pvc)]; exists {
+		glog.V(5).Infof("PersistentVolumeClaim[%s/%s] has already been provisioned.  No volume will be creatd.", pvc.Namespace, pvc.Name)
+		return nil
+	}
+
+	qos, exists := pvc.Annotations[qosProvisioningKey]
+	if !exists {
+		return fmt.Errorf("PersistentVolumeClaim[%s/%s] missing quality of service annotation", pvc.Namespace, pvc.Name)
+	}
+	glog.V(5).Infof("Provisioning PersistentVolumeClaim[%s/%s]\n", pvc.Namespace, pvc.Name)
+
+	volumeOptions := volume.VolumeOptions{
+		CapacityMB:                    100,
+		AccessModes:                   pvc.Spec.AccessModes,
+		PersistentVolumeReclaimPolicy: api.PersistentVolumeReclaimDelete,
+		QOSTier: qos,
+	}
+
+	creater, err := plugin.NewCreater(volumeOptions)
+	if err != nil {
+		return fmt.Errorf("Could not obtain Creater from plugin %s", plugin.Name())
+	}
+	// blocks until completion
+	pv, err := creater.Create()
+	if err != nil {
+		return fmt.Errorf("PersistentVolumeClaim[%s] failed provisioning: %+v", pvc.Name, err)
+	}
+	glog.V(5).Infof("PersistentVolume[%s] successfully created for %s/%s\n", pv.Name, pvc.Namespace, pvc.Name)
+
+	// this prevents duplicates from being provisioned in case updating the PVC fails it's API update
+	provisionedClaims[ClaimToProvisionableKey(pvc)] = true
+
+	// this is a persistent record that the PVC has been provisioned that is also used to prevent dupes
+	pvc.Annotations[provisionedKey] = "true"
+
+	_, err = client.UpdatePersistentVolumeClaim(pvc)
+	if err != nil {
+		// this error is bad but not fatal because we will attempt storing the claim name in the PV's annotations
+		glog.Errorf("PersistentVolumeClaim[%s] failed update.  Its provisioned annotation was not persisteed.")
+	}
+
+	pv.Annotations[createdForKey] = ClaimToProvisionableKey(pvc)
+	pv, err = client.CreatePersistentVolume(pv)
+	if err != nil {
+		// the volume was created in the infrastructure but we failed to create a PV pointing to it.
+		// the volume is now orphaned and there is no easy means of finding/recovering it.
+		// TODO:  https://github.com/kubernetes/kubernetes/issues/14443
+		return fmt.Errorf("Error creating PersistentVolume for claim %s: %+v", pvc.Name, err)
+	}
+
+	return nil
+}
+
+// Run starts this provisioner's control loops
+func (provisioner *PersistentVolumeProvisioner) Run() {
+	glog.V(5).Infof("Starting PersistentVolumeProvisioner\n")
+	if provisioner.stopChannel == nil {
+		provisioner.stopChannel = make(chan struct{})
+		go provisioner.pvclaimController.Run(provisioner.stopChannel)
```

Comment:
this needs to be like the binder and handle two stop channels for the two controllers.